### ISSUE TITLE
Provide an open API for place types

### ISSUE THREAD

**Issue by msurkovsky** [spirali/kaira#7]
Currently there is supposed that types which are used as place types provides some methods. However, there is no interface which I could use when writing a new type. It is true that there are some methods which can be put into header file of Kaira program, but when I write a new code I would like to keep it all at one place.

**Comment by spirali** [comment]
You mean some template wizard?

**Comment by msurkovsky** [comment]
I mean to extract packers functionality and maybe prepare an abstract type for place types which would define required functions.

**Comment by spirali** [comment]
I do not see a point. Each type needs own serialization functions. So having an interface can bring some better error messages (but only for internal packers) but does not save you any code.

**Comment by msurkovsky** [comment]
Yes,  I did not say it saves you any code. I don not like the situation when I use a type on a place and depending on a specific situation it can end with an error. Practically in this moment all types used on places needs `token_name` method. Then when are sent data through net it is needed packing functionality. The same holds for verification as well as new testing functionality. Therefore, I suggest to clearly say all types used on places has to be of a specific type. If they won't their usability is quite limited. In my opinion, the work with place types then would be more clear and as you said it also will bring better error messages.

**Comment by spirali** [comment]
It is not always possible, the user does not necessary have control over the used classes. E.g. you cannot change ancestor class for Octave types, or other external libraries that is not under your control.

**Comment by msurkovsky** [comment]
Well, but you can always create a new type which put original type and place interface together.  Maybe it would be the best when I try to prepare a prototype, then we can discuss above some code. I'm not 100% sure if we understand each other now :-).

**Comment by spirali** [comment]
A wrapper class will cause a extra copying or some smart-pointer semantics has to be involved. I would like to see a prototype, but I have to warn you that it is not simple problem.

**Comment by msurkovsky** [comment]
Ok, I'll try to write a prototype. Maybe I don't see something write now and I'll change my opinion. But now, I still think it's not so hard. I'm aware it won't be trivial but I don't think it would be, let say, one month project.
